### PR TITLE
Simplify additional script loading

### PR DESCRIPTION
**Pull request by nanaya**
Use jquery's existing functionality (and Promise) to load the scripts and use set instead of object for recording loaded scripts.

Discovered during exploring way to preload react components ahead of time.

### CODE REVIEW

**Review by notbakaneko — COMMENTED** [review]
Should it abort loading on navigation?

File: resources/assets/lib/turbolinks-reload.ts [hunk 1]
```diff
@@ -1,49 +1,36 @@
...
   constructor() {
-    document.addEventListener('turbolinks:before-cache', this.cleanup);
+    $(document).on('turbolinks:before-cache', this.abortLoading);
   }
 
-  cleanup = () => {
-    $(`.${className}`).remove();
+  abortLoading = () => {
+    for (const xhr of this.loading.values()) {
+      xhr.abort();
```
Comment (notbakaneko) [inline resources/assets/lib/turbolinks-reload.ts:14]: Probably should also should remove from the set after aborting

File: resources/assets/lib/turbolinks-reload.ts [hunk 2]
```diff
@@ -1,49 +1,36 @@
...
+    if (this.loaded.has(src)) return;
 
-    const el = document.createElement('script');
-    el.classList.add(className);
-    el.onload = () => {
-      // abort if the element has been removed (on navigation etc)
-      if (el.parentElement == null) {
-        return;
-      }
+    const xhr = $.ajax(src, { dataType: 'script' }) as JQuery.jqXHR<void>;
```
Comment (notbakaneko) [inline resources/assets/lib/turbolinks-reload.ts]: add `cache: true` to avoid weird `?_=number` querystring

**Review by notbakaneko — CHANGES_REQUESTED** [review]
Probably should also handle/prevent/abort multiple calls to load same script (original only loads once)

File: resources/assets/lib/turbolinks-reload.ts (hunk 1 above)
Comment (nanaya) [reply resources/assets/lib/turbolinks-reload.ts:14]: should already be done by the always callback

Comment (nanaya) [reply resources/assets/lib/turbolinks-reload.ts:14]: ...and shouldn't be in loaded set if not done

File: resources/assets/lib/turbolinks-reload.ts [hunk 3]
```diff
@@ -1,49 +1,36 @@
...
+    this.loaded.delete(src);
+    this.loading.get(src)?.abort();
   };
 
-  load = (src: string, onload?: () => void) => {
-    if (this.loaded[src]) {
-      return;
-    }
+  load(src: string) {
+    if (this.loaded.has(src) || this.loading.has(src)) return;
```
Comment (notbakaneko) [inline resources/assets/lib/turbolinks-reload.ts:24]: return the existing xhr request?

File: resources/assets/coffee/_classes/twitch-player.coffee [hunk 4]
```diff
@@ -9,7 +9,9 @@ class @TwitchPlayer
 
 
   initializeEmbed: =>
-    @turbolinksReload.load 'https://player.twitch.tv/js/embed/v1.js', @startAll
+    @turbolinksReload
+      .load 'https://player.twitch.tv/js/embed/v1.js'
+      .then @startAll
```
Comment (notbakaneko) [inline resources/assets/coffee/_classes/twitch-player.coffee:14]: `load` can be `undefined` by definition even though checking if `Twitch` is set should prevent it. Coming to think of it, wouldn't this loop infinitely if `Twitch` isn't set by the script? 🤔

**Review by notbakaneko — CHANGES_REQUESTED** [review]

File: resources/assets/coffee/_classes/twitch-player.coffee (hunk 4 above)
Comment (nanaya) [reply resources/assets/coffee/_classes/twitch-player.coffee]: the `load` wouldn't do anything if the script is already loaded

**Review by notbakaneko — APPROVED** [review]

**Pull request merged**